dvc/repo/__init__.py

```diff
@@ -360,26 +360,14 @@ def stages(self, from_directory=None, check_dag=True):
 
         stages = []
         outs = []
-        for root, dirs, files in os.walk(str(from_directory)):
-            for fname in files:
-                path = os.path.join(root, fname)
-                if not Stage.is_stage_file(path):
-                    continue
-                stage = Stage.load(self, path)
-                for out in stage.outs:
-                    outs.append(out.path + out.sep)
-                stages.append(stage)
 
-            def filter_dirs(dname, root=root):
```

Comment:
@efiop it seems obsolete, can you verify if I am right on this one?

Reply:
@pared It is not obsolete. Here we are excluding directories beforehand, so we don't spend time walking through them. It is very important when you have giant output directories.

Reply:
This would be changed in #1709. There will be a filesystem abstraction to unify the the file tree access for the files inside the specific SCM branch and the working directory. I think `dvc.utils.walk_files` instead could be made obsolete, but not sure as it looks like it is used only in places where it is only meaningful to work with the working tree and the mentioned abstraction is not needed. Duplication of `os.walk` behaviour in different is not good thing to have, though, this is the point behind of possibility to obsolete `dvc.utils.walk_files`.

Reply:
Btw, there is a problem with this `filter_dirs` implementation, it actually could make the things slower if there are many outs, its current complexity is `O(num_dirs * num_outs)` and could be reduced by using `set` to store outs and checking that it contains path or any of it parents.

Reply:
Added issue for that - #1751

Reply:
Also, this functionality could moved inside the `walk` function, but I'm not sure yet about this.